File: catalog/backends/java-quarkus/patterns/FeatureFlagStringVariant.java
package com.example.service;

import dev.openfeature.sdk.*;
import jakarta.enterprise.context.ApplicationScoped;

// Pattern: Feature Flags — string variant for A/B branching
// Adapt: replace "ui-theme" with your flag key; variant values must match Datadog flag config
@ApplicationScoped
public class FeatureFlagStringVariant {

    public String getUiVariant(String userId) {
        Client client = OpenFeatureAPI.getInstance().getClient(); // provider initialized in FeatureFlagBooleanGate
        EvaluationContext ctx = new MutableContext(userId);
        String variant = client.getStringValue("ui-theme", "control", ctx);
        return switch (variant) {
            case "treatment-a" -> "new_design_a";
            case "treatment-b" -> "new_design_b";
            default -> "control_design";
        };
    }
}

File: catalog/backends/go-gin/patterns/feature_flag_string_variant.go
package patterns

import (
	"context"

	"github.com/open-feature/go-sdk/openfeature"
)

// Pattern: Feature Flags — string variant for A/B branching
// Adapt: replace "ui-theme" with your flag key; variant values must match Datadog flag config
// Uses flagClient initialized in feature_flag_boolean_gate.go

var variants = map[string]string{
	"treatment-a": "new_design_a",
	"treatment-b": "new_design_b",
}

func GetUiVariant(ctx context.Context, userID string) string {
	evalCtx := openfeature.NewEvaluationContext(userID, nil)
	variant, _ := flagClient.StringValue(ctx, "ui-theme", "control", evalCtx)
	if result, ok := variants[variant]; ok {
		return result
	}
	return "control_design"
}
